### PR TITLE
discv5 support

### PR DESCRIPTION
**Pull request by supernovahs**
#1383 

TODO 
Add relevant traits in https://github.com/sigp/discv5 through a PR

### CODE REVIEW

**Comment by mattsse** [comment]
> Add relevant traits in https://github.com/sigp/discv5 through a PR

this should not be required, everything we need on top of discv5 should go into the reth-discv5 crate

File: crates/net/network/src/discovery.rs [hunk 1]
```diff
@@ -77,6 +105,57 @@ impl Discovery {
...
+            let discv5_clone_for_async = Arc::clone(&discv5);
+
+            let discv5_service_handle = runtime.spawn(async move {
+
+                let discv5_locked = discv5_clone_for_async.lock().unwrap();
+
+                if let Err(err) = discv5_locked.start().await {
+                    println!("Discv5 start error: {:?}", err);
+                }
+            });
```
Comment (supernovahs) [inline crates/net/network/src/discovery.rs:152]: @mattsse discv5 does not implement clone trait . How should I go with this ?

```
future cannot be sent between threads safely
within `[async block@crates/net/network/src/discovery.rs:145:55: 152:14]`, the trait `std::marker::Send` is not implemented for `std::sync::MutexGuard<'_, Discv5>`

```

File: crates/net/network/src/discovery.rs [hunk 2]
```diff
@@ -77,6 +105,57 @@ impl Discovery {
...
+            let discv5 = Arc::new(Mutex::new(Discv5::new(enr, enr_key, discv5_config).unwrap()));
+            let discv5_ = Arc::clone(&discv5);
+            let mut service = discv5_.lock().unwrap();
+
+            runtime.block_on(service.start()).map_err(|err| {
+                NetworkError::from_discv5_error(
+                    err.to_string(),
+                    ServiceKind::Discovery(discovery_addr),
+                )
+            })?;
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:130]: I don't understand why this is necessary.

**Review by mattsse — CHANGES_REQUESTED** [review]
okay after looking at an example:

https://github.com/sigp/discv5/blob/aa12e384c52cc706a570b24260fef09f41c0181b/examples/find_nodes.rs#L159

I think all of this should be easier than expected, we perhaps don't even need the additional crate since the start function spawns the services.

what we should do here is

1. Configure discv5
2. `call start`
3. get an event stream
4. Store `Arc<Discv5>` internally as field and the eventstream as field

**Comment by supernovahs** [comment]
`lint / MSRV / ethereum (pull_request) ` wants me to import enr crate , but clippy does not want me to add (its redundant).  😅

**Comment by supernovahs** [comment]
@rkrasiuk let me know what additional changes need to be done.
Thanks

**Review by mattsse — CHANGES_REQUESTED** [review]
great!

what's missing now is polling the discv5 event stream in the poll function

File: crates/net/discv5/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    default::Default,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc;
+use tokio_stream::{wrappers::ReceiverStream, Stream};
+
+fn allow_all_enrs(_enr: &Enr) -> bool {
+    true
+}
```
Comment (rkrasiuk) [inline crates/net/discv5/src/lib.rs:16]:
```suggestion

/// The default table filter that results in all nodes being accepted into the local routing table.
const fn allow_all_enrs(_enr: &Enr) -> bool {
    true
}
```

File: crates/net/network/src/discovery.rs [hunk 4]
```diff
@@ -51,6 +57,25 @@ pub struct Discovery {
...
 }
 
+impl fmt::Debug for Discovery {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Discovery")
+            .field("discovered_nodes", &self.discovered_nodes)
+            .field("local_enr", &self.local_enr)
+            .field("discv4", &self.discv4)
+            // Use a placeholder for `discv5` since it doesn't implement Debug
+            .field("discv5", &format_args!("<Discv5>"))
```
[the remark this answers is not in the document]
Comment (supernovahs) [reply crates/net/network/src/discovery.rs:67]: cool

File: crates/net/network/Cargo.toml [hunk 5]
```diff
@@ -31,7 +32,7 @@ reth-rpc-types.workspace = true
 reth-tokio-util.workspace = true
 
 alloy-rlp.workspace = true
-
+k256 = "0.13.2"
```
[the remark this answers is not in the document]
Comment (supernovahs) [reply crates/net/network/Cargo.toml:35]: yes 
https://github.com/supernovahs/reth/blob/965f58ded5ef11cb8e6cf0041fc04559ece3bad1/crates/net/network/src/discovery.rs#L8

File: crates/net/network/src/discovery.rs [hunk 6]
```diff
@@ -105,16 +87,32 @@ impl Discovery {
...
+
+                let enr_key = CombinedKey::Secp256k1(signing_key);
+                let enr = enr::EnrBuilder::new("v4")
+                    .build(&enr_key)
+                    .map_err(|e| Discv5Error::EnrBuilderConstruct(e.to_string()))?;
+
+                Discv5Wrapper::new(
+                    Discv5::new(enr, enr_key, discv5_config)
+                        .map_err(|e| Discv5Error::Discv5Construct(e.to_string()))?,
+                )
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:103]: all of this should be moved to Discv5Wrapper

File: crates/net/network/src/discovery.rs [hunk 7]
```diff
@@ -232,6 +230,17 @@ impl Discovery {
         }
     }
 
+    fn on_discv5_update(&mut self, update: Event) {
+        match update {
+            Event::Discovered(enr) => {}
+            Event::EnrAdded { enr, replaced } => {}
+            Event::NodeInserted { node_id, replaced } => {}
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:237]: here we need to handle the discovered entries

see https://github.com/paradigmxyz/reth/blob/32bf97d7bc7a862e5440f3182428de427585bb0f/crates/net/network/src/discovery.rs#L169-L169

File: crates/net/discv5/src/lib.rs [hunk 8]
```diff
@@ -5,16 +5,80 @@ pub use discv5::{
...
+    SecretKeyDecode(String),
+    /// Failed to construct a new EnrBuilder
+    #[error("failed to constuct a new EnrBuilder {0}")]
+    EnrBuilderConstruct(String),
+    /// Failed to construct Discv5 instance
+    #[error("failed to construct a new Discv5 instance {0}")]
+    Discv5Construct(String),
+    /// Failed to create a event stream
+    #[error("failed to create event stream {0}")]
+    Discv5EventStream(String),
```
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:76]: these should use `#[from]` where possible or retain the wrapped error type

File: crates/net/discv5/src/lib.rs [hunk 9]
```diff
@@ -5,16 +5,80 @@ pub use discv5::{
...
+    sync::Arc,
     task::{Context, Poll},
 };
-use tokio::sync::mpsc;
+use tokio::sync::{mpsc, Mutex, MutexGuard};
 use tokio_stream::{wrappers::ReceiverStream, Stream};
 
-fn allow_all_enrs(_enr: &Enr) -> bool {
+// Wrapper struct for Discv5
+pub struct Discv5Wrapper {
```
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:17]:
```suggestion
pub struct Discv5Handle {
```

File: crates/net/discv5/src/lib.rs [hunk 10]
```diff
@@ -5,16 +5,80 @@ pub use discv5::{
 use futures_util::StreamExt;
 use std::{
     default::Default,
+    fmt,
     pin::Pin,
+    sync::Arc,
     task::{Context, Poll},
 };
-use tokio::sync::mpsc;
+use tokio::sync::{mpsc, Mutex, MutexGuard};
```
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:13]: we want a parking_lo::mutex here

**Review by mattsse — CHANGES_REQUESTED** [review]
pretty close

File: crates/net/network/src/discovery.rs [hunk 11]
```diff
@@ -188,6 +215,80 @@ impl Discovery {
...
+                    id: node_id,
+                };
+                self.on_node_record_update(node_record, None)
+            }
+            Event::SocketUpdated(socket_address) => {
+                if let Some(discv5) = &self.discv5 {
+                    let discv5 = Discv5Handle::convert_to_discv5(discv5);
+                    let discv5_guard = discv5.lock();
+                    let mut local_enr = discv5_guard.local_enr();
+                    // local_enr.set_ip(socket_address.ip(),); #5576 figure out how to get signing
```
Comment (supernovahs) [inline crates/net/network/src/discovery.rs:274]: trying to figure out how to get the signing key to update the ip

**Comment by supernovahs** [comment]
Some errors need to be handled apart from open comments.
Also, I am reviewing the implementation again to check for any bugs

**Comment by supernovahs** [comment]
@mattsse any inputs on how to fix the test error. It occurred after using parking lot mutex . 
I tried debugging it. 
thanks

File: crates/net/discv5/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::{sync::mpsc, task};
+use tokio_stream::{wrappers::ReceiverStream, Stream};
+
+// Wrapper struct for Discv5
+#[derive(Clone)]
+pub struct Discv5Handle {
+    inner: Arc<Mutex<Discv5>>,
```
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:23]: why do we need a mutex here?

File: crates/net/discv5/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+            Discv5::new(enr, enr_key, discv5_config)
+                .map_err(|_e| Discv5Error::Discv5Construct.into())?,
+        ))
+    }
+
+    pub fn convert_to_discv5(&self) -> Arc<Mutex<Discv5>> {
+        self.inner.clone()
+    }
+
+    pub async fn start_service(&self) -> Result<(), Discv5Error> {
```
Comment (mattsse) [inline crates/net/discv5/src/lib.rs:53]: can this not call start directly on the wrapped discv5?

File: crates/net/network/src/discovery.rs [hunk 14]
```diff
@@ -77,6 +85,21 @@ impl Discovery {
...
+        // setup discv5
+        let (discv5,) = if let Some(discv5_config) = discv5_config {
+            let discv5 = Discv5Handle::from_secret_key(sk, discv5_config)?;
+            discv5.clone().start_service().await.unwrap();
+            // let discv5_event_stream : ;
+            // let discv5_event = discv5
+            //     .create_event_stream()
+            //     .await
+            //     .map_err(|_e| NetworkError::Discv5(Discv5Error::SecretKeyDecode))?;
+            // let discv5_event_stream = Some(ReceiverStream::new(discv5_event));
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:97]: please remove

File: crates/net/network/src/discovery.rs [hunk 15]
```diff
@@ -188,6 +213,80 @@ impl Discovery {
...
+            Event::TalkRequest(talk_request) => {
+                if let Some(discv5) = &self.discv5 {
+                    let discv5 = Discv5Handle::convert_to_discv5(discv5);
+                    let discv5_guard = discv5.lock();
+                    let node_id = talk_request.node_id();
+                    let enr = discv5_guard.find_enr(&node_id);
+                    let protocol = talk_request.protocol();
+                    let request = talk_request.body();
+                    let _ =
+                        discv5_guard.talk_req(enr.unwrap(), protocol.to_vec(), request.to_vec());
```
Comment (mattsse) [inline crates/net/network/src/discovery.rs:284]: I don't think we need this

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by gakonst** [comment]
@supernovahs anything else you'd like to add? else @mattsse bump

**Comment by supernovahs** [comment]
closing since #6151  is now the main PR for this

**Pull request closed without merging**